### PR TITLE
EOS-24754 : Added Put-bucket-replication API related files

### PR DESCRIPTION
**Pull request by shraddhaghatol**
Signed-off-by: shraddhaghatol <shraddha.j.ghatol@seagate.com>

# Problem Statement
Implementation of Put-bucket-replication API

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: server/s3_put_bucket_replication_action.cc [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  s3_log(S3_LOG_INFO, stripped_request_id, "Consume data\n");
+  if (request->is_s3_client_read_error()) {
+    client_read_error();
+  } else if (request->has_all_body_content()) {
+    new_bucket_replication_content = request->get_full_body_content_as_string();
+    validate_request_body(new_bucket_replication_content);
+  } else {
+    // else just wait till entire body arrives. rare.
+    request->resume();
+  }
```
Comment (knrajnambiar76) [inline server/s3_put_bucket_replication_action.cc:83]: Have exit log

File: server/s3_put_replication_body.cc [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        s3_log(S3_LOG_WARN, request_id, "XML request body Invalid.\n");
+      }
+    }
+    
+    rule_child_node = rule_child_node->next;
+  } while (rule_child_node != nullptr);
+
+  if (is_destination_node_present == false) {
+    s3_log(S3_LOG_WARN, request_id,
+           "XML request body Invalid.Destination node not present\n");
```
Comment (mbcortx) [inline server/s3_put_replication_body.cc:322]: checking whether destination field is given or not, correct?

File: server/s3_put_bucket_replication_action.cc (hunk 1 above)
Comment (shraddhaghatol) [reply server/s3_put_bucket_replication_action.cc:83]: Added Exit log

File: server/s3_put_replication_body.cc (hunk 2 above)
Comment (shraddhaghatol) [reply server/s3_put_replication_body.cc:473]: yes.Correct.

File: resources/s3_error_messages.json [hunk 3]
```diff
@@ -194,5 +194,29 @@
   "MissingContentLength": {
     "Description": "You must provide the Content-Length HTTP header.",
     "httpcode": 411
+  },
+  "InvalidArgumentDestinationBucket": {
+    "Description": "IDestination bucket cannot be null or empty.",
```
Comment (mbcortx) [inline resources/s3_error_messages.json:199]: IDestination->Destination

**Review by mbcortx — APPROVED** [review]
LGTM

File: resources/s3_error_messages.json (hunk 3 above)
Comment (shraddhaghatol) [reply resources/s3_error_messages.json]: Done changes

File: server/s3_object_metadata.cc [hunk 4]
```diff
@@ -872,6 +872,11 @@ void S3ObjectMetadata::set_tags(
   object_tags = tags_as_map;
 }
 
+void S3ObjectMetadata::set_bucket_replication_configuration(
```
Comment (rajkumarpatel2602) [inline server/s3_object_metadata.cc:875]: check this function and correct if necessary

File: server/s3_object_metadata.h [hunk 5]
```diff
@@ -63,6 +63,7 @@ class S3ObjectMetadataCopyable {
   std::map<std::string, std::string> system_defined_attribute;
   std::map<std::string, std::string> user_defined_attribute;
   std::map<std::string, std::string> object_tags;
+  std::string replication_config;
```
Comment (rajkumarpatel2602) [inline server/s3_object_metadata.h:66]: should be inside bucket metadata

File: server/s3_put_bucket_replication_action.cc [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+void S3PutBucketReplicationAction::setup_steps() {
+  s3_log(S3_LOG_DEBUG, request_id, "Setting up the action\n");
+  ACTION_TASK_ADD(S3PutBucketReplicationAction::validate_request, this);
+  ACTION_TASK_ADD(
+      S3PutBucketReplicationAction::save_replication_config_to_bucket_metadata,
+      this);
+  ACTION_TASK_ADD(S3PutBucketReplicationAction::send_response_to_s3_client,
+                  this);
+  // ...
```
Comment (rajkumarpatel2602) [inline server/s3_put_bucket_replication_action.cc:53]: can be removed

File: server/s3_put_bucket_replication_action.cc [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+           "Bucket metadata load operation failed due to pre launch failure\n");
+    set_s3_error("ServiceUnavailable");
+  } else {
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void
```
Comment (rajkumarpatel2602) [inline server/s3_put_bucket_replication_action.cc:150]: make it consistent with other function heading style. if formatter doesn't help then need to convey to other s3 team.

File: server/s3_put_bucket_replication_action.h [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#pragma once
```
Comment (rajkumarpatel2602) [inline server/s3_put_bucket_replication_action.h:21]: can be placed inside #ifdef check for __S3_SERVER_S3_PUT_BUCKET_REPLICATION_ACTION_H__.

File: server/s3_put_replication_body.h [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+#ifndef __S3_SERVER_S3_PUT_REPLICATION_BODY_H__
+#define __S3_SERVER_S3_PUT_REPLICATION_BODY_H__
+
+#include <json/json.h>
+#include <gtest/gtest_prod.h>
+#include <string>
+#include <map>
+#include <libxml/xmlmemory.h>
+
+// #define BUCKET_MAX_TAGS 50
```
Comment (rajkumarpatel2602) [inline server/s3_put_replication_body.h:32]: can be removed if not required.

File: server/s3_object_metadata.cc (hunk 4 above)
Comment (shraddhaghatol) [reply server/s3_object_metadata.cc]: Removed the function.

File: server/s3_object_metadata.h (hunk 5 above)
Comment (shraddhaghatol) [reply server/s3_object_metadata.h]: removed the declaration.

File: server/s3_put_bucket_replication_action.cc (hunk 6 above)
Comment (shraddhaghatol) [reply server/s3_put_bucket_replication_action.cc]: Done.

File: server/s3_put_bucket_replication_action.cc (hunk 7 above)
Comment (shraddhaghatol) [reply server/s3_put_bucket_replication_action.cc:149]: We format the code using clang-format -style and if the function name is big, while formatting using command - clang-format -style=Google -i it does this automatically. If we try to adjust it, it gives formatting error.

File: server/s3_put_bucket_replication_action.h (hunk 8 above)
Comment (shraddhaghatol) [reply server/s3_put_bucket_replication_action.h:21]: #pragma is used to ensure the header file gets included only once.
Referred other APIs files, they are having same format. i.e outside #ifdef

File: server/s3_put_replication_body.h (hunk 9 above)
Comment (shraddhaghatol) [reply server/s3_put_replication_body.h]: done

File: server/s3_put_bucket_replication_action.h (hunk 8 above)
Comment (rajkumarpatel2602) [reply server/s3_put_bucket_replication_action.h:21]: Actually #ifdef check also does similar thing. No problem.

**Comment by johnbent** [comment]
Two approvals. All comments addressed. Merging.

**Pull request merged**